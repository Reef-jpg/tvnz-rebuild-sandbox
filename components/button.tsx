import styled from "styled-components";

const StyledButton = styled.button`
  max-height: 2.5rem;
  height: 2.5rem;
  padding: 0.75rem 1.1875rem;

  display: flex;
  justify-content: center;
  align-items: center;
  gap: 11px;

  cursor: pointer;
  border-radius: 8px;
  border: 0;

  /* font styling */
  font-family: "Black Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 12px;
  letter-spacing: 0.875px;
  text-transform: uppercase;

  transition: all 0.2s ease-in-out;

  .icon {
    line-height: 0;

    img {
      height: 100%;
    }
  }

  &:hover {
    box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.1);
  }

  /* variant styling */
  &.primary {
    background-color: #0778d6;

    &:hover {
      background-color: #054e8c;
    }
  }
  &.secondary {
    background-color: transparent;
    color: #f5f5f5;
  }
  &.outline {
    border: 1px solid white;
    background-color: transparent;

    &:hover {
      background-color: rgba(255, 255, 255, 0.5);
    }
  }

  /* circle styling */
  &.circle {
    border-radius: 9001px;
    padding: 1rem;
    max-height: 50px;
    max-width: 50px;
    height: 50px;
    width: 50px;
  }

  /* my list animations */
  &.in-my-list {
    .icon {
      animation-name: squeeze1;
      animation-duration: 0.6s;
      animation-iteration-count: 1;
      animation-delay: 0.01s;
    }
    @keyframes squeeze1 {
      50% {
        transform: scale(0);
      }

      100% {
        transform: scale(1);
      }
    }
  }
  &.not-in-my-list {
    .icon {
      animation-name: squeeze2;
      animation-duration: 0.6s;
      animation-iteration-count: 1;
      animation-delay: 0.01s;
    }
    @keyframes squeeze2 {
      50% {
        transform: scale(0);
      }

      100% {
        transform: scale(1);
      }
    }
  }
`;

// ts types for button variations
type ButtonVariant =
  | "primary"
  | "secondary"
  | "outline"
  | "primary circle"
  | "secondary circle"
  | "outline circle";

const Button = ({
  text = "",
  icon = "",
  onClick,
  isSelected = false,
  variant = "primary",
  className,
}: {
  text?: string;
  icon?: string;
  onClick?: () => void;
  isSelected?: boolean;
  variant?: ButtonVariant;
  className?: string;
}) => {
  return (
    <StyledButton
      onClick={onClick}
      className={`button ${isSelected ? "selected" : ""} 
        ${variant} ${className}`}
    >
      {/* icon */}
      {icon && (
        <span className="icon">
          <img src={icon} alt="icon" />
        </span>
      )}
      {/* text */}
      {text && <span className="text">{text}</span>}
    </StyledButton>
  );
};

export default Button;
