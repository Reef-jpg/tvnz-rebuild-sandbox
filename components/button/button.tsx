import { MouseEventHandler } from "react";
import styled from "styled-components";

const StyledButton = styled.button`
  padding: 0.8rem 1.1875rem;

  display: flex;
  justify-content: center;
  align-items: center;
  gap: 11px;

  cursor: pointer;
  border-radius: 8px;
  border: 0;

  /* font styling */
  font-family: "Black Sans";
  font-style: normal;
  font-weight: 700;
  font-size: 0.75rem;
  letter-spacing: 0.875px;
  text-transform: uppercase;

  transition: all 0.2s ease-in-out;

  .icon {
    line-height: 0;

    img {
      height: 100%;
    }
  }

  &:hover {
    box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.1);
  }

  /* variant styling */
  &.primary {
    background-color: #0778d6;

    &:hover {
      background-color: #054e8c;
    }
  }
  &.secondary {
    background-color: transparent;
    color: #f5f5f5;
  }
  &.outline {
    border: 1px solid white;
    background-color: transparent;

    &:hover {
      background-color: rgba(255, 255, 255, 0.5);
    }
  }

  /* circle styling */
  &.circle {
    border-radius: 9001px;
    padding: 1rem;
    max-height: 3.125rem;
    max-width: 3.125rem;
    height: 3.125rem;
    width: 3.125rem;
  }

  /* my list animations */
  &.in-my-list {
    .icon {
      height: 16px;
      width: 16px;
      transform: scale(0.7);
      animation-name: squeeze1;
      animation-duration: 0.6s;
      animation-iteration-count: 1;
      animation-delay: 0.01s;
    }
    @keyframes squeeze1 {
      50% {
        transform: scale(0);
      }

      100% {
        transform: scale(1);
      }
    }
  }
  &.not-in-my-list {
    .icon {
      height: 16px;
      width: 16px;
      animation-name: squeeze2;
      animation-duration: 0.6s;
      animation-iteration-count: 1;
      animation-delay: 0.01s;
    }
    @keyframes squeeze2 {
      50% {
        transform: scale(0);
      }

      100% {
        transform: scale(1);
      }
    }
  }
`;

// ts types for button variations
type ButtonVariant =
  | "primary"
  | "secondary"
  | "outline"
  | "primary circle"
  | "secondary circle"
  | "outline circle";

export const Button = ({
  text = "",
  icon = "",
  iconAlternativeText = "Button Icon",
  onClick,
  isSelected = false,
  variant = "primary",
  className = "",
}: {
  text?: string;
  icon?: string;
  iconAlternativeText?: string;
  onClick?: MouseEventHandler;
  isSelected?: boolean;
  variant?: ButtonVariant;
  className?: string;
}) => {
  return (
    <StyledButton
      onClick={onClick}
      className={`button ${isSelected ? "selected" : ""} 
        ${variant} ${className}`}
    >
      {/* icon */}
      {icon && (
        <span className="icon">
          <img src={icon} alt={iconAlternativeText} />
        </span>
      )}
      {/* text */}
      {text && <span className="text">{text}</span>}
    </StyledButton>
  );
};
