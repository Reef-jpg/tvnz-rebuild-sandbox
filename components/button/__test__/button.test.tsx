import "@testing-library/jest-dom";

import renderer, { act } from "react-test-renderer";

import { Button } from "../button";
import { render } from "@testing-library/react";

describe("Button Component", () => {
  test("test defaults", () => {
    const { container } = render(<Button />);

    expect(container.querySelector(".text")).toBeFalsy();
    expect(container.querySelector(".icon")).toBeFalsy();
    expect(container.querySelector(".primary")).toBeTruthy();
  });

  test("should have text", () => {
    const { getByText } = render(<Button text="test" />);

    expect(getByText("test")).toBeInTheDocument();
  });

  test("should have image and alt text", () => {
    const { getByAltText } = render(
      <Button icon="/test.svg" iconAlt="Test Button Alt" />
    );

    expect(getByAltText("Test Button Alt")).toBeInTheDocument();
    expect(getByAltText("Test Button Alt")).toHaveAttribute("src", "/test.svg");
  });

  test("should have classNames", () => {
    const { getByRole } = render(
      <Button variant="secondary circle" className="test" isSelected={true} />
    );

    expect(getByRole("button")).toHaveClass("test");
    expect(getByRole("button")).toHaveClass("secondary");
    expect(getByRole("button")).toHaveClass("circle");
    expect(getByRole("button")).toHaveClass("selected");
  });

  test("should be no undefined classNames", () => {
    const { getByText, container } = render(
      <Button text="undefined test" icon="/undefined.svg" />
    );

    expect(getByText("undefined test").closest("button")).not.toHaveClass(
      "undefined"
    );
    expect(container.querySelector(".text")).not.toHaveClass("undefined");
    expect(container.querySelector(".icon")).not.toHaveClass("undefined");
  });

  test("onclick handler should work", () => {
    const mock = jest.fn();
    const { getByRole } = render(<Button onClick={mock} />);

    act(() => {
      getByRole("button").click();
      getByRole("button").click();
      getByRole("button").click();
    });
    expect(mock).toBeCalledTimes(3);
  });

  test("should match snapshots", () => {
    const withProps = renderer
      .create(
        <Button
          text="Test"
          icon="/test.svg"
          iconAlt="test icon alt text"
          variant="secondary"
          className="test"
          isSelected={true}
          onClick={() => console.log("test clicked")}
        />
      )
      .toJSON();

    const withoutProps = renderer.create(<Button />).toJSON();

    expect(withProps).toMatchSnapshot();
    expect(withoutProps).toMatchSnapshot();
  });
});
