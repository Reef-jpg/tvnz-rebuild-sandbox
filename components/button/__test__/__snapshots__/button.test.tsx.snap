// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should match snapshots 1`] = `
<button
  className="button__StyledButton-sc-1476jkj-0 bXdlHz button selected 
        secondary test"
  onClick={[Function]}
>
  <span
    className="icon"
  >
    <img
      alt="test icon alt text"
      src="/test.svg"
    />
  </span>
  <span
    className="text"
  >
    Test
  </span>
</button>
`;

exports[`Button Component should match snapshots 2`] = `
<button
  className="button__StyledButton-sc-1476jkj-0 bXdlHz button  
        primary "
/>
`;
